### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by Annwan**
Adding documentation where it is lacking all over the library

### CODE REVIEW

**Comment by aquarial** [comment]
Looks nice! 

This shows up in the hackage documentation right?

Good add to the light docs on the readme for those who go through hackage.

**Comment by Annwan** [comment]
Yeah that goes in the hackage doc, but also shows up when you use a project-local hoogle instance (e.g. run via `stack hoogle`), which is my main reason for doing those

**Comment by Annwan** [comment]
Just rebased on the latest commits from master

**Comment by Annwan** [comment]
That should be everything document-able documented
`Discord.Internal.Gateway.Cache` maybe lacks a bit but I don't really understand how it works and the names looked plenty explicit

File: src/Discord/Internal/Types/Channel.hs [hunk 1]
```diff
@@ -25,111 +25,121 @@ import Discord.Internal.Types.Emoji
...
-      , channelPosition    :: Integer
-      , channelNSFW        :: Bool
-      , channelPermissions :: [Overwrite]
-      , channelParentId    :: Maybe ParentId
+      { channelId          :: ChannelId      -- ^ The id of the channel
+      , channelGuild       :: GuildId        -- ^ The id of the guild
+      , channelName        :: T.Text         -- ^ The name of the channel (2 - 1000 characters)
+      , channelPosition    :: Integer        -- ^ The position of the channel
+      , channelNSFW        :: Bool           -- ^ Is not-safe-for-work
+      , channelPermissions :: [Overwrite]    -- ^ An array of permission 'Overrite's
```
Comment (aquarial) [inline src/Discord/Internal/Types/Channel.hs:60]: The formatting here is nice

File: src/Discord/Internal/Types/Gateway.hs [hunk 2]
```diff
@@ -51,7 +54,7 @@ data GatewayIntent = GatewayIntent
   , gatewayIntentWebhooks :: Bool
   , gatewayIntentInvites :: Bool
   , gatewayIntentVoiceStates :: Bool
-  , gatewayIntentPrecenses :: Bool
+  , gatewayIntentPresences :: Bool
```
Comment (aquarial) [inline src/Discord/Internal/Types/Gateway.hs:57]: Haha I never noticed this, good spellcheck.

Might be a minor breaking change, but will generate a very nice compiler warning (`similar variable in scope: did you mean [...]`) so it'll be fine to release

**Review by aquarial — APPROVED** [review]
Impressive work! Fills in so many documentation holes. 

And the `BatchEditApplicationCommandPermissions` removal is a gem.

**Pull request merged**